File: ios/AmazonIVSBroadcastModule.swift
import Foundation
import React
import AmazonIVSBroadcast
import AVFoundation

@objc(AmazonIVSBroadcastModule)
class AmazonIVSBroadcastModule: RCTEventEmitter, IVSBroadcastSession.Delegate {
  override static func moduleName() -> String! {
    return "AmazonIVSBroadcastModule"
  }

  override static func requiresMainQueueSetup() -> Bool {
    return true
  }

  static var sharedSession: IVSBroadcastSession?

  private var broadcastSession: IVSBroadcastSession?
  private var currentCamera: IVSDevice?
  private var currentMicrophone: IVSDevice?
  private var hasListeners = false

  // MARK: - EventEmitter
  override func supportedEvents() -> [String]! {
    return [
      "stateChanged",
      "error",
      "deviceChanged",
      "bitrateChanged",
      "networkHealthChanged",
      "networkQualityChanged",
      "audioSessionInterrupted",
      "audioSessionResumed",
      "cameraError",
      "microphoneError",
      "reconnecting",
      "reconnected"
    ]
  }

  override func startObserving() {
    hasListeners = true
  }

  override func stopObserving() {
    hasListeners = false
  }

  private func sendEvent(_ name: String, body: Any?) {
    if hasListeners {
      sendEvent(withName: name, body: body)
    }
  }

  // MARK: - IVSBroadcastSession.Delegate
  func broadcastSession(_ session: IVSBroadcastSession, didChange state: IVSBroadcastSession.State) {
    sendEvent("stateChanged", body: ["state": state.rawValue])
  }

  func broadcastSession(_ session: IVSBroadcastSession, didEmitError error: Error) {
    let nsError = error as NSError
    let isFatal = (error as NSError).userInfo[IVSBroadcastErrorIsFatalKey] as? Bool ?? false
    sendEvent("error", body: [
      "message": error.localizedDescription,
      "code": nsError.domain + ":" + String(nsError.code),
      "isFatal": isFatal
    ])
  }

  func broadcastSession(_ session: IVSBroadcastSession, didChange devices: [IVSDevice]) {
    let deviceList = devices.map { device -> [String: Any] in
      return [
        "id": device.descriptor().uid,
        "name": device.descriptor().name,
        "position": device.descriptor().position?.rawValue ?? "",
        "type": device.type.rawValue
      ]
    }
    sendEvent("deviceChanged", body: deviceList)
  }

  func broadcastSession(_ session: IVSBroadcastSession, didChangeBitrate bitrate: Int) {
    sendEvent("bitrateChanged", body: ["bitrate": bitrate])
  }

  func broadcastSession(_ session: IVSBroadcastSession, didChangeNetworkHealth health: IVSTransmissionStatistics.NetworkHealth) {
    sendEvent("networkHealthChanged", body: ["networkHealth": health.rawValue])
  }

  func broadcastSession(_ session: IVSBroadcastSession, audioSessionWasInterrupted interrupted: Bool) {
    sendEvent(interrupted ? "audioSessionInterrupted" : "audioSessionResumed", body: nil)
  }

  func broadcastSessionDidStartReconnect(_ session: IVSBroadcastSession) {
    sendEvent("reconnecting", body: nil)
  }

  func broadcastSessionDidReconnect(_ session: IVSBroadcastSession) {
    sendEvent("reconnected", body: nil)
  }

  // MARK: - Yayın Yönetimi
  @objc(startBroadcast:withResolver:withRejecter:)
  func startBroadcast(options: NSDictionary, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    guard let rtmpsUrl = options["rtmpsUrl"] as? String,
          let streamKey = options["streamKey"] as? String else {
      reject("E_MISSING_PARAMS", "rtmpsUrl ve streamKey gereklidir", nil)
      return
    }

    do {
      let config = try createBroadcastConfig(from: options)
      let devices = try createDeviceDescriptors(from: options)
      
      self.broadcastSession = try IVSBroadcastSession(
        configuration: config,
        descriptors: devices,
        delegate: self
      )
      
      // Auto-reconnect ayarları
      if let enableAutoReconnect = options["enableAutoReconnect"] as? Bool {
        self.broadcastSession?.shouldAutoReconnect = enableAutoReconnect
        
        if let maxRetries = options["autoReconnectMaxRetries"] as? Int {
          self.broadcastSession?.maxReconnectAttempts = maxRetries
        }
        
        if let retryInterval = options["autoReconnectRetryInterval"] as? Double {
          self.broadcastSession?.reconnectRetryInterval = retryInterval
        }
      }
      
      AmazonIVSBroadcastModule.sharedSession = self.broadcastSession
      
      try self.broadcastSession?.start(with: rtmpsUrl, streamKey: streamKey)
      resolve(nil)
    } catch let error {
      reject("E_BROADCAST_START", "Yayın başlatılamadı: \(error.localizedDescription)", error)
    }
  }

  private func createBroadcastConfig(from options: NSDictionary) throws -> IVSBroadcastConfiguration {
    let config = IVSPresets.configurations().standardLandscape()
    
    // Video ayarları
    if let videoConfig = options["video"] as? [String: Any] {
      if let width = videoConfig["width"] as? Int,
         let height = videoConfig["height"] as? Int {
        try config.video.setSize(CGSize(width: width, height: height))
      }
      if let bitrate = videoConfig["bitrate"] as? Int {
        try config.video.setInitialBitrate(bitrate)
      }
      if let targetFramerate = videoConfig["targetFramerate"] as? Int {
        try config.video.setTargetFramerate(targetFramerate)
      }
      if let keyframeInterval = videoConfig["keyframeInterval"] as? Int {
        try config.video.setKeyframeInterval(keyframeInterval)
      }
      if let maxBitrate = videoConfig["maxBitrate"] as? Int {
        try config.video.setMaxBitrate(maxBitrate)
      }
      if let minBitrate = videoConfig["minBitrate"] as? Int {
        try config.video.setMinBitrate(minBitrate)
      }
      if let qualityOptimization = videoConfig["qualityOptimization"] as? String {
        config.video.qualityOptimization = qualityOptimization == "quality" ? .quality : .latency
      }
      if let useH265 = videoConfig["useH265"] as? Bool {
        config.video.useH265 = useH265
      }
      if let enableTranscoding = videoConfig["enableTranscoding"] as? Bool {
        config.video.enableTranscoding = enableTranscoding
      }
    }
    
    // Audio ayarları
    if let audioConfig = options["audio"] as? [String: Any] {
      if let bitrate = audioConfig["bitrate"] as? Int {
        try config.audio.setBitrate(bitrate)
      }
      if let channels = audioConfig["channels"] as? Int {
        try config.audio.setChannels(channels)
      }
      if let sampleRate = audioConfig["sampleRate"] as? Int {
        try config.audio.setSampleRate(sampleRate)
      }
      if let enableEchoCancellation = audioConfig["enableEchoCancellation"] as? Bool {
        config.audio.enableEchoCancellation = enableEchoCancellation
      }
      if let enableNoiseSuppression = audioConfig["enableNoiseSuppression"] as? Bool {
        config.audio.enableNoiseSuppression = enableNoiseSuppression
      }
    }
    
    // Mixer ayarları
    if let mixerConfig = options["mixer"] as? [String: Any] {
      if let canvasWidth = mixerConfig["canvasWidth"] as? Int,
         let canvasHeight = mixerConfig["canvasHeight"] as? Int {
        config.mixer.setCanvasSize(CGSize(width: canvasWidth, height: canvasHeight))
      }
      if let backgroundColor = mixerConfig["backgroundColor"] as? String {
        config.mixer.backgroundColor = UIColor(hexString: backgroundColor)
      }
    }

    return config
  }

  private func createDeviceDescriptors(from options: NSDictionary) -> [IVSDeviceDescriptor] {
    var descriptors: [IVSDeviceDescriptor] = []
    
    // Kamera ayarları
    if let cameraPosition = options["cameraPosition"] as? String {
      let position: IVSDeviceDescriptor.Position = cameraPosition == "back" ? .back : .front
      descriptors.append(contentsOf: IVSPresets.devices().camera(position: position))
    } else {
      descriptors.append(contentsOf: IVSPresets.devices().frontCamera())
    }
    
    // Mikrofon ayarları
    descriptors.append(contentsOf: IVSPresets.devices().microphone())
    
    return descriptors
  }

  @objc(stopBroadcast:withRejecter:)
  func stopBroadcast(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    guard let session = self.broadcastSession else {
      reject("E_NO_SESSION", "Aktif bir yayın yok", nil)
      return
    }
    
    session.stop()
    self.broadcastSession = nil
    AmazonIVSBroadcastModule.sharedSession = nil
    resolve(nil)
  }

  // MARK: - Cihaz Yönetimi
  @objc(switchCamera:withRejecter:)
  func switchCamera(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    guard let session = self.broadcastSession else {
      reject("E_NO_SESSION", "Aktif bir yayın yok", nil)
      return
    }

    let cameras = session.listAvailableDevices().filter { $0.type == .camera }
    guard let current = self.currentCamera,
          let currentIndex = cameras.firstIndex(where: { $0.descriptor().uid == current.descriptor().uid }) else {
      reject("E_NO_CAMERA", "Mevcut kamera bulunamadı", nil)
      return
    }

    let nextIndex = (currentIndex + 1) % cameras.count
    let nextCamera = cameras[nextIndex]

    session.exchangeOldDevice(current, withNewDevice: nextCamera) { [weak self] newDevice, error in
      if let error = error {
        reject("E_SWITCH_CAMERA", "Kamera değiştirilemedi: \(error.localizedDescription)", error)
      } else if let device = newDevice {
        self?.currentCamera = device
        resolve(["position": device.descriptor().position?.rawValue ?? "unknown"])
      }
    }
  }

  @objc(setMicrophoneMuted:withResolver:withRejecter:)
  func setMicrophoneMuted(muted: Bool, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    guard let session = self.broadcastSession else {
      reject("E_NO_SESSION", "Aktif bir yayın yok", nil)
      return
    }

    guard let mic = currentMicrophone as? IVSMicrophone else {
      reject("E_NO_MIC", "Aktif mikrofon bulunamadı", nil)
      return
    }

    mic.setMuted(muted)
    resolve(nil)
  }

  // MARK: - Kalite ve İstatistikler
  @objc(getTransmissionStatistics:withRejecter:)
  func getTransmissionStatistics(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    guard let session = self.broadcastSession else {
      reject("E_NO_SESSION", "Aktif bir yayın yok", nil)
      return
    }

    let stats = session.transmissionStatistics
    resolve([
      "currentBitrate": stats.currentBitrate,
      "recommendedBitrate": stats.recommendedBitrate,
      "roundTripTime": stats.roundTripTime,
      "networkHealth": stats.networkHealth.rawValue,
      "broadcastQuality": stats.broadcastQuality.rawValue
    ])
  }

  // MARK: - Yeni Özellikler
  @objc(getStreamMetrics:withRejecter:)
  func getStreamMetrics(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    guard let session = self.broadcastSession else {
      reject("E_NO_SESSION", "Aktif bir yayın yok", nil)
      return
    }

    let metrics = session.streamMetrics
    resolve([
      "cpu": metrics.cpuUsage,
      "memory": metrics.memoryUsage,
      "battery": metrics.batteryLevel,
      "temperature": metrics.deviceTemperature
    ])
  }

  @objc(configureAudioSession:withResolver:withRejecter:)
  func configureAudioSession(config: NSDictionary, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    guard let session = self.broadcastSession else {
      reject("E_NO_SESSION", "Aktif bir yayın yok", nil)
      return
    }

    do {
      let category = config["category"] as? String ?? "playAndRecord"
      let mode = config["mode"] as? String ?? "default"
      let mixWithOthers = config["mixWithOthers"] as? Bool ?? false

      try session.configureAudioSession(
        category: audioSessionCategory(from: category),
        mode: audioSessionMode(from: mode),
        options: mixWithOthers ? .mixWithOthers : []
      )
      resolve(nil)
    } catch let error {
      reject("E_AUDIO_SESSION", "Audio session yapılandırılamadı: \(error.localizedDescription)", error)
    }
  }

  private func audioSessionCategory(from string: String) -> AVAudioSession.Category {
    switch string {
    case "ambient": return .ambient
    case "playback": return .playback
    case "record": return .record
    case "playAndRecord": return .playAndRecord
    default: return .playAndRecord
    }
  }

  private func audioSessionMode(from string: String) -> AVAudioSession.Mode {
    switch string {
    case "voiceChat": return .voiceChat
    case "gameChat": return .gameChat
    case "videoChat": return .videoChat
    default: return .default
    }
  }
} 

File: android/src/main/java/com/amazonivs/AmazonIVSBroadcastModule.java
package com.amazonivs;

import android.media.AudioManager;
import android.util.Log;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import androidx.annotation.NonNull;

import com.amazonaws.ivs.broadcast.BroadcastConfiguration;
import com.amazonaws.ivs.broadcast.BroadcastConfiguration.AspectMode;
import com.amazonaws.ivs.broadcast.BroadcastConfiguration.Video;
import com.amazonaws.ivs.broadcast.BroadcastConfiguration.Audio;
import com.amazonaws.ivs.broadcast.BroadcastConfiguration.Mixer;
import com.amazonaws.ivs.broadcast.BroadcastSession;
import com.amazonaws.ivs.broadcast.BroadcastSession.State;
import com.amazonaws.ivs.broadcast.BroadcastSession.Listener;
import com.amazonaws.ivs.broadcast.Device;
import com.amazonaws.ivs.broadcast.DeviceDiscovery;
import com.amazonaws.ivs.broadcast.ImageDevice;
import com.amazonaws.ivs.broadcast.Microphone;
import com.amazonaws.ivs.broadcast.TransmissionStats;

import java.util.List;

public class AmazonIVSBroadcastModule extends ReactContextBaseJavaModule {
    public static BroadcastSession sharedSession;
    private BroadcastSession broadcastSession;
    private DeviceDiscovery deviceDiscovery;
    private Device currentCamera;
    private Device currentMicrophone;
    private final ReactApplicationContext reactContext;

    public AmazonIVSBroadcastModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
        this.deviceDiscovery = new DeviceDiscovery(reactContext);
    }

    @NonNull
    @Override
    public String getName() {
        return "AmazonIVSBroadcastModule";
    }

    private void sendEvent(String eventName, Object params) {
        reactContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
            .emit(eventName, params);
    }

    private final Listener sessionListener = new Listener() {
        @Override
        public void onStateChanged(@NonNull BroadcastSession session, @NonNull State state) {
            WritableMap map = Arguments.createMap();
            map.putString("state", state.name());
            sendEvent("stateChanged", map);
        }

        @Override
        public void onError(@NonNull BroadcastSession session, @NonNull Exception e) {
            WritableMap map = Arguments.createMap();
            map.putString("message", e.getMessage());
            map.putString("code", e.getClass().getSimpleName());
            map.putBoolean("isFatal", e instanceof BroadcastSession.FatalError);
            sendEvent("error", map);
        }

        @Override
        public void onDeviceChange(@NonNull BroadcastSession session, @NonNull List<Device> devices) {
            WritableArray arr = Arguments.createArray();
            for (Device d : devices) {
                WritableMap dev = Arguments.createMap();
                dev.putString("id", d.getDescriptor().getUid());
                dev.putString("name", d.getDescriptor().getName());
                dev.putString("position", d.getDescriptor().getPosition().name());
                dev.putString("type", d.getType().name());
                arr.pushMap(dev);
            }
            sendEvent("deviceChanged", arr);
        }

        @Override
        public void onBitrateChanged(@NonNull BroadcastSession session, int bitrate) {
            WritableMap map = Arguments.createMap();
            map.putInt("bitrate", bitrate);
            sendEvent("bitrateChanged", map);
        }

        @Override
        public void onNetworkHealthChanged(@NonNull BroadcastSession session, @NonNull TransmissionStats.NetworkHealth health) {
            WritableMap map = Arguments.createMap();
            map.putString("networkHealth", health.name());
            sendEvent("networkHealthChanged", map);
        }

        @Override
        public void onAudioSessionInterrupted(@NonNull BroadcastSession session) {
            sendEvent("audioSessionInterrupted", null);
        }

        @Override
        public void onAudioSessionResumed(@NonNull BroadcastSession session) {
            sendEvent("audioSessionResumed", null);
        }

        @Override
        public void onReconnectStart(@NonNull BroadcastSession session) {
            sendEvent("reconnecting", null);
        }

        @Override
        public void onReconnectSuccess(@NonNull BroadcastSession session) {
            sendEvent("reconnected", null);
        }
    };

    @ReactMethod
    public void startBroadcast(ReadableMap options, Promise promise) {
        if (broadcastSession != null) {
            promise.reject("E_ALREADY_RUNNING", "Yayın zaten başlatılmış.");
            return;
        }
        if (!options.hasKey("rtmpsUrl") || !options.hasKey("streamKey")) {
            promise.reject("E_MISSING_PARAMS", "rtmpsUrl ve streamKey gereklidir");
            return;
        }

        String rtmpsUrl = options.getString("rtmpsUrl");
        String streamKey = options.getString("streamKey");

        try {
            BroadcastConfiguration config = createBroadcastConfig(options);
            broadcastSession = new BroadcastSession(reactContext, deviceDiscovery, config, sessionListener);
            sharedSession = broadcastSession;
            broadcastSession.start(rtmpsUrl, streamKey);
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("E_BROADCAST_START", "Yayın başlatılamadı: " + e.getMessage(), e);
        }
    }

    private BroadcastConfiguration createBroadcastConfig(ReadableMap options) {
        BroadcastConfiguration config = new BroadcastConfiguration();

        // Video ayarları
        if (options.hasKey("video")) {
            ReadableMap videoConfig = options.getMap("video");
            Video video = config.video;
            
            if (videoConfig.hasKey("width") && videoConfig.hasKey("height")) {
                video.setSize(
                    videoConfig.getInt("width"),
                    videoConfig.getInt("height")
                );
            }
            
            if (videoConfig.hasKey("bitrate")) {
                video.setInitialBitrate(videoConfig.getInt("bitrate"));
            }
            
            if (videoConfig.hasKey("targetFramerate")) {
                video.setTargetFramerate(videoConfig.getInt("targetFramerate"));
            }
            
            if (videoConfig.hasKey("keyframeInterval")) {
                video.setKeyframeInterval(videoConfig.getInt("keyframeInterval"));
            }

            if (videoConfig.hasKey("maxBitrate")) {
                video.setMaxBitrate(videoConfig.getInt("maxBitrate"));
            }

            if (videoConfig.hasKey("minBitrate")) {
                video.setMinBitrate(videoConfig.getInt("minBitrate"));
            }

            if (videoConfig.hasKey("qualityOptimization")) {
                String optimization = videoConfig.getString("qualityOptimization");
                video.setQualityOptimization(
                    optimization.equals("quality") ? 
                    Video.QualityOptimization.QUALITY : 
                    Video.QualityOptimization.LATENCY
                );
            }

            if (videoConfig.hasKey("useH265")) {
                video.setUseH265(videoConfig.getBoolean("useH265"));
            }

            if (videoConfig.hasKey("enableTranscoding")) {
                video.setEnableTranscoding(videoConfig.getBoolean("enableTranscoding"));
            }
        }

        // Audio ayarları
        if (options.hasKey("audio")) {
            ReadableMap audioConfig = options.getMap("audio");
            Audio audio = config.audio;
            
            if (audioConfig.hasKey("bitrate")) {
                audio.setBitrate(audioConfig.getInt("bitrate"));
            }
            
            if (audioConfig.hasKey("channels")) {
                audio.setChannels(audioConfig.getInt("channels"));
            }

            if (audioConfig.hasKey("sampleRate")) {
                audio.setSampleRate(audioConfig.getInt("sampleRate"));
            }

            if (audioConfig.hasKey("enableEchoCancellation")) {
                audio.setEnableEchoCancellation(audioConfig.getBoolean("enableEchoCancellation"));
            }

            if (audioConfig.hasKey("enableNoiseSuppression")) {
                audio.setEnableNoiseSuppression(audioConfig.getBoolean("enableNoiseSuppression"));
            }
        }

        // Mixer ayarları
        if (options.hasKey("mixer")) {
            ReadableMap mixerConfig = options.getMap("mixer");
            Mixer mixer = config.mixer;
            
            if (mixerConfig.hasKey("canvasWidth") && mixerConfig.hasKey("canvasHeight")) {
                mixer.setCanvasSize(
                    mixerConfig.getInt("canvasWidth"),
                    mixerConfig.getInt("canvasHeight")
                );
            }

            if (mixerConfig.hasKey("backgroundColor")) {
                String color = mixerConfig.getString("backgroundColor");
                mixer.setBackgroundColor(Color.parseColor(color));
            }
        }

        // Auto-reconnect ayarları
        if (options.hasKey("enableAutoReconnect")) {
            config.setEnableAutoReconnect(options.getBoolean("enableAutoReconnect"));

            if (options.hasKey("autoReconnectMaxRetries")) {
                config.setAutoReconnectMaxRetries(options.getInt("autoReconnectMaxRetries"));
            }

            if (options.hasKey("autoReconnectRetryInterval")) {
                config.setAutoReconnectRetryInterval(options.getDouble("autoReconnectRetryInterval"));
            }
        }

        return config;
    }

    @ReactMethod
    public void stopBroadcast(Promise promise) {
        if (broadcastSession == null) {
            promise.reject("E_NO_SESSION", "Aktif bir yayın yok");
            return;
        }
        
        broadcastSession.stop();
        broadcastSession = null;
        sharedSession = null;
        promise.resolve(null);
    }

    @ReactMethod
    public void switchCamera(Promise promise) {
        if (broadcastSession == null) {
            promise.reject("E_NO_SESSION", "Aktif bir yayın yok");
            return;
        }

        List<Device> cameras = deviceDiscovery.listAvailableDevices(Device.Type.CAMERA);
        if (currentCamera == null && !cameras.isEmpty()) {
            currentCamera = cameras.get(0);
        }
        if (cameras.size() < 2) {
            promise.reject("E_NO_CAMERA", "Birden fazla kamera yok");
            return;
        }

        int idx = cameras.indexOf(currentCamera);
        int nextIdx = (idx + 1) % cameras.size();
        Device nextCamera = cameras.get(nextIdx);

        broadcastSession.exchangeOldDevice(currentCamera, nextCamera, (newDevice, error) -> {
            if (error != null) {
                promise.reject("E_SWITCH_CAMERA", "Kamera değiştirilemedi: " + error.getMessage(), error);
            } else {
                currentCamera = newDevice;
                WritableMap result = Arguments.createMap();
                result.putString("position", newDevice.getDescriptor().getPosition().name());
                promise.resolve(result);
            }
        });
    }

    @ReactMethod
    public void setMicrophoneMuted(boolean muted, Promise promise) {
        if (broadcastSession == null) {
            promise.reject("E_NO_SESSION", "Aktif bir yayın yok");
            return;
        }

        if (currentMicrophone == null || !(currentMicrophone instanceof Microphone)) {
            promise.reject("E_NO_MIC", "Aktif mikrofon bulunamadı");
            return;
        }

        ((Microphone) currentMicrophone).setMuted(muted);
        promise.resolve(null);
    }

    @ReactMethod
    public void getTransmissionStatistics(Promise promise) {
        if (broadcastSession == null) {
            promise.reject("E_NO_SESSION", "Aktif bir yayın yok");
            return;
        }

        TransmissionStats stats = broadcastSession.getTransmissionStatistics();
        WritableMap result = Arguments.createMap();
        result.putInt("currentBitrate", stats.getCurrentBitrate());
        result.putInt("recommendedBitrate", stats.getRecommendedBitrate());
        result.putInt("roundTripTime", stats.getRoundTripTime());
        result.putString("networkHealth", stats.getNetworkHealth().name());
        result.putString("broadcastQuality", stats.getBroadcastQuality().name());
        promise.resolve(result);
    }

    @ReactMethod
    public void getStreamMetrics(Promise promise) {
        if (broadcastSession == null) {
            promise.reject("E_NO_SESSION", "Aktif bir yayın yok");
            return;
        }

        WritableMap metrics = Arguments.createMap();
        metrics.putDouble("cpu", broadcastSession.getStreamMetrics().getCpuUsage());
        metrics.putDouble("memory", broadcastSession.getStreamMetrics().getMemoryUsage());
        metrics.putDouble("battery", broadcastSession.getStreamMetrics().getBatteryLevel());
        metrics.putDouble("temperature", broadcastSession.getStreamMetrics().getDeviceTemperature());
        promise.resolve(metrics);
    }

    @ReactMethod
    public void configureAudioSession(ReadableMap config, Promise promise) {
        if (broadcastSession == null) {
            promise.reject("E_NO_SESSION", "Aktif bir yayın yok");
            return;
        }

        try {
            String category = config.hasKey("category") ? config.getString("category") : "playAndRecord";
            String mode = config.hasKey("mode") ? config.getString("mode") : "default";
            boolean mixWithOthers = config.hasKey("mixWithOthers") ? config.getBoolean("mixWithOthers") : false;

            AudioManager.Mode audioMode = getAudioMode(mode);
            int streamType = getStreamType(category);

            AudioManager audioManager = (AudioManager) reactContext.getSystemService(Context.AUDIO_SERVICE);
            audioManager.setMode(audioMode);
            audioManager.setStreamVolume(streamType, audioManager.getStreamMaxVolume(streamType), 0);
            
            if (mixWithOthers) {
                audioManager.setMicrophoneMute(false);
                audioManager.setSpeakerphoneOn(true);
            }

            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("E_AUDIO_SESSION", "Audio session yapılandırılamadı: " + e.getMessage());
        }
    }

    private int getStreamType(String category) {
        switch (category) {
            case "ambient":
                return AudioManager.STREAM_SYSTEM;
            case "playback":
                return AudioManager.STREAM_MUSIC;
            case "record":
                return AudioManager.STREAM_VOICE_CALL;
            case "playAndRecord":
            default:
                return AudioManager.STREAM_VOICE_CALL;
        }
    }

    private int getAudioMode(String mode) {
        switch (mode) {
            case "voiceChat":
                return AudioManager.MODE_IN_COMMUNICATION;
            case "gameChat":
                return AudioManager.MODE_NORMAL;
            case "videoChat":
                return AudioManager.MODE_IN_COMMUNICATION;
            default:
                return AudioManager.MODE_NORMAL;
        }
    }
} 